instrumentor/controllers/instrumentedapplication_controller.go

```diff
@@ -0,0 +1,362 @@
+	pod, err := r.choosePods(ctx, labels, app.Namespace)
+	if err != nil {
+		return err
+	}
+
+	langDetectionPod, err := r.createLangDetectionPod(pod, app)
+	if err != nil {
+		return err
+	}
+
+	err = r.Create(ctx, langDetectionPod)
+	return err
+}
+
+func (r *InstrumentedApplicationReconciler) choosePods(ctx context.Context, labels map[string]string, namespace string) (*corev1.Pod, error) {
+	var podList corev1.PodList
+	err := r.List(ctx, &podList, client.MatchingLabels(labels), client.InNamespace(namespace))
+	if err != nil {
+		return nil, err
+	}
+
+	if len(podList.Items) == 0 {
+		return nil, consts.PodsNotFoundErr
+	}
+
+	for _, pod := range podList.Items {
+		if pod.Status.Phase == corev1.PodRunning {
+			return &pod, nil
+		}
+	}
+
+	return nil, consts.PodsNotFoundErr
+}
+
+func (r *InstrumentedApplicationReconciler) createLangDetectionPod(targetPod *corev1.Pod, instrumentedApp *v1.InstrumentedApplication) (*corev1.Pod, error) {
+	pod := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: fmt.Sprintf("%s-instrumentation-detection-", targetPod.Name),
+			Namespace:    targetPod.Namespace,
+			Annotations: map[string]string{
+				consts.InstrumentationDetectionContainerAnnotationKey: "true",
+				istioAnnotationKey:   istioAnnotationValue,
+				linkerdAnnotationKey: linkerdAnnotationValue,
+			},
+		},
+		Spec: corev1.PodSpec{
+			Containers: []corev1.Container{
+				{
+					Name:  "instrumentation-detector",
+					Image: fmt.Sprintf("%s:%s", r.InstrumentationDetectorImage, r.InstrumentationDetectorTag),
+					Args: []string{
+						fmt.Sprintf("--pod-uid=%s", targetPod.UID),
+						fmt.Sprintf("--container-names=%s", strings.Join(r.getContainerNames(targetPod), ",")),
+					},
+					TerminationMessagePath: "/dev/detection-result",
+					SecurityContext: &corev1.SecurityContext{
+						Capabilities: &corev1.Capabilities{
+							Add: []corev1.Capability{"SYS_PTRACE"},
+						},
+					},
+				},
+			},
+			RestartPolicy: "Never",
+			NodeName:      targetPod.Spec.NodeName,
+			HostPID:       true,
+		},
+	}
+
+	err := ctrl.SetControllerReference(instrumentedApp, pod, r.Scheme)
+	if err != nil {
+		return nil, err
+	}
+
+	return pod, nil
+}
+
+func (r *InstrumentedApplicationReconciler) getContainerNames(pod *corev1.Pod) []string {
+	var result []string
+	for _, c := range pod.Spec.Containers {
+		if !r.skipContainer(c.Name) {
+			result = append(result, c.Name)
+		}
+	}
+
+	return result
+}
+
+func (r *InstrumentedApplicationReconciler) skipContainer(name string) bool {
+	return name == "istio-proxy" || name == "linkerd-proxy"
+}
+
+func (r *InstrumentedApplicationReconciler) getOwnerTemplateLabels(ctx context.Context, instrumentedApp *v1.InstrumentedApplication) (map[string]string, error) {
+	owner := metav1.GetControllerOf(instrumentedApp)
+	if owner == nil {
+		return nil, errors.New("could not find owner for InstrumentedApp")
+	}
+
+	if owner.Kind == "Deployment" && owner.APIVersion == appsv1.SchemeGroupVersion.String() {
```

Comment:
Maybe add to `common.go` consts like:
supportedResourceDeployment = "Deployment" and use them instead of explicit strings?